hw30/index.js

```diff
@@ -0,0 +1,223 @@
+            if (this.isSelected) {
+                this.elem.classList.add('selected');
+            } else {
+                this.elem.classList.remove('selected');
+            }
+        }
+
+        setElement(element) {
+            this.elem = element;
+        }
+
+        remove() {
+            this.element.remove();
+        }
+
+        get element() {
+            return this.elem;
+        }
+    }
+
+    class List {
+        items = [];
+
+        constructor() {
+            const list = document.createElement('ol');
+            this.element = list;
+            this.#registerEvents(this.element);
+        }
+
+        #registerEvents(element) {
+            document.addEventListener('click', (e) => {
+                if (!e.target.closest('.container')) {
+                    this.clearSelected();
+                }
+            })
+            element.addEventListener('mousedown', function (e) {
+                e.preventDefault();
+            });
+
+            element.addEventListener('click', (e) => {
+
+                if (e.target === element) {
+                    return false;
+                }
+
+                if (e.metaKey || e.ctrlKey) {
+                    const selected = this.getListItem(e.target);
+                    const newState = !selected.isSelected;
+                    selected.setSelected(newState);
+                    return;
+                }
+
+                if (!e.shiftKey) {
+                    this.clearSelected();
+                }
+
+                if (e.shiftKey) {
+                    const currentlySelected = this.getSelected();
+
+                    if (!currentlySelected) {
+                        const children = this.items;
+                        for (let index = 0; index < children.length; index++) {
+                            const listItem = children[index];
+                            listItem.setSelected(true);
+                            if (e.target === listItem.element) {
+                                break;
+                            }
+                        }
+                        return;
+                    }
+
+                    const targetIndex = this.items.findIndex((listItem) => listItem.element === e.target);
+                    const currentlySelectedIndex = this.items.findIndex((listItem) => listItem.element === currentlySelected[0].element);
+                    this.clearSelected();
+                    const [start, end] = targetIndex < currentlySelectedIndex ? [targetIndex, currentlySelectedIndex] : [currentlySelectedIndex, targetIndex];
+
+                    for (let index = start; index <= end; index++) {
+                        this.items[index].setSelected(true);
+                    }
+                    return;
+                }
+
+                this.clearSelected();
+                const selected = this.getListItem(e.target);
+                selected.setSelected(true);
+            });
```

Comment:
You have strange structure of this function, its really hard to read.First you have logic of exiting from function - it's ok. But then you have a conditional branching, which really hard to read.
you need to do something like:
```javascript
if (e.ctrlKey || e.metaKey) {
 // ctrlKey behaviour 
} else if (e.shiftKey) {
 // shiftKey behaviour
} else {
   // any other behaviour
}
```

Reply:
Also shiftKey behaviour need to be wrapped in a method (`selectRange` for example). And it can be refactored to use inly one loop (hint: this loop is enough ```for (let index = start; index <= end; index++)```)